examples/pipeline_wavernn/datasets.py

```diff
@@ -0,0 +1,142 @@
+import os
+import random
+
+import torch
+import torchaudio
+from torchaudio.datasets import LJSPEECH
+
+from transform import linear_to_mel
+from functional import (
+    label_to_waveform,
+    mulaw_encode,
+    specgram_normalize,
+    waveform_to_label,
+)
+
+
+class MapMemoryCache(torch.utils.data.Dataset):
+    r"""Wrap a dataset so that, whenever a new item is returned, it is saved to memory.
+    """
+
+    def __init__(self, dataset):
+        self.dataset = dataset
+        self._cache = [None] * len(dataset)
+
+    def __getitem__(self, n):
+        if self._cache[n] is not None:
+            return self._cache[n]
+
+        item = self.dataset[n]
+        self._cache[n] = item
+
+        return item
+
+    def __len__(self):
+        return len(self.dataset)
+
+
+class ProcessedLJSPEECH(LJSPEECH):
+    def __init__(self, dataset, transforms, args):
+        self.dataset = dataset
+        self.transforms = transforms
+        self.args = args
+
+    def __getitem__(self, index):
+        filename = self.dataset[index][0]
+        folder = "LJSpeech-1.1/wavs/"
+        file = os.path.join(self.args.file_path, folder, filename + ".wav")
+
+        return self.process_datapoint(file)
+
+    def __len__(self):
+        return len(self.dataset)
+
+    def process_datapoint(self, file):
+        args = self.args
+        n_fft = 2048
+        waveform, sample_rate = torchaudio.load(file)
+        specgram = self.transforms(waveform)
+        # TODO Replace by torchaudio, once https://github.com/pytorch/audio/pull/593 is resolved
+        specgram = linear_to_mel(specgram, sample_rate, n_fft, args.n_freq, args.f_min)
+        specgram = specgram_normalize(specgram, args.min_level_db)
+        waveform = waveform.squeeze(0)
+
+        if args.mode == "waveform":
+            waveform = (
+                mulaw_encode(waveform, 2 ** args.n_bits)
+                if args.mulaw
+                else waveform_to_label(waveform, args.n_bits)
+            )
+
+        return waveform, specgram
+
+
+def split_data(data, val_ratio, seed):
+    dataset = data._walker
+
+    random.seed(seed)
+    random.shuffle(dataset)
+
+    train_dataset = dataset[: -int(val_ratio * len(dataset))]
+    val_dataset = dataset[-int(val_ratio * len(dataset)):]
+
+    return train_dataset, val_dataset
+
+
+def gen_datasets_ljspeech(
+    args, transforms,
+):
+    data = LJSPEECH(root=args.file_path, download=False)
+
+    train_dataset, val_dataset = split_data(data, args.val_ratio, args.seed)
+
+    train_dataset = ProcessedLJSPEECH(train_dataset, transforms, args)
+    val_dataset = ProcessedLJSPEECH(val_dataset, transforms, args)
+
+    train_dataset = MapMemoryCache(train_dataset)
+    val_dataset = MapMemoryCache(val_dataset)
+
+    return train_dataset, val_dataset
```

Comment:
Maybe something like this instead? Thoughts? I haven't tested :)

```python
def gen_datasets_ljspeech(
    args, transforms,
):

    data = LJSPEECH(root=args.file_path, download=False)
    train_dataset, val_dataset = split_data(data, args.val_ratio, args.seed)

    train_dataset = Processed(train_dataset, transforms)
    val_dataset = Processed(val_dataset, transforms)

    train_dataset = MapMemoryCache(train_dataset)
    val_dataset = MapMemoryCache(val_dataset)
    return train_dataset, val_dataset

```

```python
class Splitted(torch.utils.data.Dataset):
    def __init__(self, dataset, indices):
        self._dataset = dataset
        self._indices = indices
        super().__init__()

    def __getitem__(self, key):
        return self._dataset[self._indices[key]]

def split_data(data, val_ratio, seed):

    random.seed(seed)
    train_indices = random.sample(range(len(data)), k = int(val_ratio * len(data)))
    val_indices = [i for i in range(len(data)) if i not in train_indices]  # does not need to be shuffled

    return Splitted(data, train_indices), Splitted(data, val_indices)
```

```python
class Processed(torch.utils.data.Dataset):
    def __init__(self, dataset, transforms):
        self._dataset = dataset
        self.transforms = transforms
        super().__init__()

    def __getitem__(self, key):
        item = self._dataset[key]
        return self.process_datapoint(item)

    def process_datapoint(self, item):
        # ...
        return transformed, target

    def __len__(self):
        return len(self._dataset)
```

Reply:
See also [this](https://discuss.pytorch.org/t/torch-utils-data-dataset-random-split/32209/4)

Reply:
Thanks for updating this :)